Fix: move additional metrics from approximator to networks

vpratz opened the pull request:
Supplying the additional metrics for inference and summary networks via the approximators compile method caused problems during deseralization (#497). This can be resolved nicely by moving the metrics directly to the networks' constructors, analogous to how Keras normally handles custom metrics in layers.

As summary networks and inference networks inherit from the respective base classes, this change only requires minor adaptations. Calls to `layer_kwargs` are now only used in classes that directly inherit from `keras.Layer`, and have been moved into `InferenceNetwork` and `SummaryNetwork`.

stefanradev93 commented:
If this is not breaking in any downstream task and only adds to the existing functionality, we may consider doing a PR to keras or asking them to implement it (or why they decided not to)?

vpratz commented:
@LarsKue Thanks for taking a look. I'll take a look at alternative ways to achieve this behavior.

vpratz commented:
@LarsKue Thanks a lot, this was a really good pointer. I have now implemented a wrapper around the constructor that is applied inside the `serializable` decorator. In addition, the default `from_config` is replaced with a variant that deserializes the config again.

There is still some copied code, but I think it is now limited to the part that is required for the feature to behave as similar to the Keras implementation as possible, apart from the desired changes (we might not need the part regarding the `dtype`, but I'm not sure and would leave it in there for now).

With those changes, we could remove the `from_config` method from most classes. Is this a change I should make?

@stefanradev93 @LarsKue Please take another look and let me know what you think (the most relevant part is in [`bayesflow/utils/serialization.py`](https://github.com/bayesflow-org/bayesflow/pull/500/files#diff-803e415eda4efd8208ec8f043e5db7a62b8409e6ac4c178b99723b8b287fbf5d)

LarsKue commented:
Thanks, this looks much better. Since this is a sensitive change, I think we should extensively test it before rolling it out. Otherwise, green light from my side!

vpratz commented:
I forgot to make the same changes in the `ModelComparisonApproximator`, this is now fixed. I have changed the output that is supplied to its classifier metrics from predictions to probabilities. This way, it is compatible with the metrics offered by Keras.

vpratz commented:
I see some weird behavior due to the inheritance, I will have to give this another look...
